apiserver/facades/agent/uniter/uniter.go

```diff
@@ -1679,6 +1680,24 @@ func (u *UniterAPIV8) WatchUnitAddresses(args params.Entities) (params.NotifyWat
 	return result, nil
 }
 
+// getUnitFallback returns a unit indirection by first checking the cache,
+// then if not found, state.
+func (u *UniterAPI) getUnitFallback(tag names.UnitTag) (unit, error) {
+	var cUnit cache.Unit
+	var unit unit
+
+	cUnit, err := u.getCacheUnit(tag)
```

Comment:
So I think we still end up with a problem of "this solves initialization, but doesn't handle charm upgrade". In the case of upgrade, we'll have marked a new charm url, but then we go to do something and we find out that we're still at the same version (because the cache hasn't been updated). Does that seem correct? (my point is just that there are ways to be out of sync in the cache that aren't just NotFound.)